lib/src/core/engine.dart

```diff
@@ -192,49 +187,61 @@ class Engine extends Disposable with EventsEmittable<EngineEvent> {
   Future<void> sendDataPacket(
     lk_models.DataPacket packet,
   ) async {
-    // make sure we do have a data connection
-    await _ensurePublisherConnected();
+    //
+    rtc.RTCDataChannel? publisherDataChannel(Reliability reliability) =>
+        reliability == Reliability.reliable ? _reliableDCPub : _lossyDCPub;
+
+    rtc.RTCDataChannelState publisherDataChannelState(
+            Reliability reliability) =>
+        publisherDataChannel(reliability)?.state ??
+        rtc.RTCDataChannelState.RTCDataChannelClosed;
+
+    final reliability = packet.kind.toSDKType();
 
     // construct the data channel message
     final message =
         rtc.RTCDataChannelMessage.fromBinary(packet.writeToBuffer());
 
-    // chose data channel
-    final rtc.RTCDataChannel? channel =
-        packet.kind == lk_models.DataPacket_Kind.LOSSY ? _lossyDC : _reliableDC;
+    if (_subscriberPrimary) {
+      // make sure publisher transport is connected
 
-    // send if channel exists
-    if (channel == null) {
-      throw UnexpectedStateException('Data channel is not ready');
-    }
+      if (publisher?.pc.iceConnectionState?.isConnected() != true) {
+        logger.fine('Publisher is not connected...');
 
-    logger.fine('sendDataPacket(label:${channel.label})');
-    await channel.send(message);
-  }
+        // start negotiation
+        if (publisher?.pc.iceConnectionState !=
+            rtc.RTCIceConnectionState.RTCIceConnectionStateChecking) {
+          await negotiate();
+        }
 
-  Future<void> _ensurePublisherConnected() async {
-    logger.fine('ensurePublisherConnected()');
-    if (!_subscriberPrimary) {
-      return;
-    }
+        logger.fine('Waiting for publisher to ice-connect...');
+        await events.waitFor<EnginePublisherIceStateUpdatedEvent>(
+          filter: (event) => event.iceState.isConnected(),
+          duration: Timeouts.iceConnection,
+        );
+      }
 
-    if (publisher?.pc.iceConnectionState?.isConnected() == true) {
-      logger.warning('[$objectId] publisher is already connected');
-      return;
+      // wait for data channel to open (if not already)
```

Comment:
This is where it waits for `open` state of the data channel.

Reply:
Is there some callback to indicate data channel open failure if it fails for some reason?

Reply:
`waitFor` will throw a `TimeoutException` if it timeouts.